src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Routing.Internal;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.AspNetCore.Routing.Template;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.ObjectPool;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class RouteValuesBasedEndpointFinder : IEndpointFinder<RouteValuesContext>
+    {
+        private readonly CompositeEndpointDataSource _endpointDataSource;
+        private readonly IInlineConstraintResolver _inlineConstraintResolver;
+        private readonly ObjectPool<UriBuildingContext> _objectPool;
+        private readonly LinkGenerationDecisionTree _linkGenerationTree;
+
+        public RouteValuesBasedEndpointFinder(
+            CompositeEndpointDataSource endpointDataSource,
+            ObjectPool<UriBuildingContext> objectPool)
+        {
+            // TODO: register for change events?
+            _endpointDataSource = endpointDataSource;
+            _objectPool = objectPool;
+
+            // TODO: Review: IInlineConstraintResolver is registered as Transient, where as we want this finder to be
+            // registered as singleton.
+            _inlineConstraintResolver = new DefaultInlineConstraintResolver(Options.Create(new RouteOptions()));
+
+            var outboundRouteEntries = GetOutboundRouteEntries();
+
+            var outboundMatches = GetOutboundMatches(outboundRouteEntries);
+
+            _linkGenerationTree = new LinkGenerationDecisionTree(outboundMatches.ToArray());
+        }
+
+        public IEnumerable<Endpoint> FindEndpoints(RouteValuesContext context)
+        {
+            var matches = _linkGenerationTree.GetMatches(context.ExplicitValues, context.AmbientValues);
+
+            return matches
+                .Select(r => r.Match.Entry)
+                .OfType<OutboundRouteEntryWithEndpoint>()
+                .Select(r => r.Endpoint);
+        }
+
+        private IList<OutboundRouteEntryWithEndpoint> GetOutboundRouteEntries()
+        {
+            var outboundRouteEntries = new List<OutboundRouteEntryWithEndpoint>();
+
+            var endpoints = _endpointDataSource.Endpoints.OfType<MatcherEndpoint>();
+            foreach (var endpoint in endpoints)
+            {
+                var requiredValuesMetadata = endpoint.Metadata.GetMetadata<IRequiredLinkValuesMetadata>();
+                var requiredLinkValues = requiredValuesMetadata?.Values;
+                if (requiredLinkValues == null)
+                {
+                    requiredLinkValues = new RouteValueDictionary();
+                }
```

Comment:
As discussed, let's make sure `MatcherEndpoint` exposes both `defaults` and `values` - where value is always a subset of defaults - defaults includes default values specified in the template and in `MapRoute(...)`